File: latest/week2/workshop4_m8/W04_M08_demo.R
# ============================================================
# Workshop 4 — Module 8: EDA Case Study — e-Car Loan Pricing
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output",showWarnings=FALSE)

ecar <- read_csv("ecar.csv") |>
  rename(CarType = `Car  Type`, CompRate = `Competition rate`,
         PrevRate = `Previous Rate`, CoF = `Cost of Funds`,
         PartnerBin = `Partner Bin`, ApproveDate = `Approve Date`) |>
  mutate(PrevRate = as.numeric(PrevRate), ApproveDate = mdy(ApproveDate),
         Year = year(ApproveDate), Spread = Rate - CoF,
         Tier = factor(Tier, levels=1:5, ordered=TRUE),
         is_new = is.na(PrevRate), CarType = str_trim(CarType))
cat("Rows:", nrow(ecar), "| Cols:", ncol(ecar), "\n")

# ── DASHBOARD ──
p_fico <- ggplot(ecar |> sample_n(20000), aes(x=FICO)) +
  geom_histogram(bins=50,fill="#1565C0",color="white") +
  geom_vline(xintercept=median(ecar$FICO),color="#E53935",linetype="dashed") +
  theme_minimal(base_size=8) + labs(title="(a) FICO")

p_rate <- ggplot(ecar |> sample_n(20000), aes(x=Rate)) +
  geom_histogram(bins=50,fill="#2E7D32",color="white") +
  theme_minimal(base_size=8) + labs(title="(b) Rate")

p_scatter <- ecar |> sample_n(5000) |>
  ggplot(aes(x=FICO,y=Rate)) + geom_point(alpha=0.05,size=0.3,color="#1565C0") +
  geom_smooth(method="lm",color="#E53935",se=FALSE) +
  theme_minimal(base_size=8) + labs(title="(c) FICO vs Rate")

p_tier <- ggplot(ecar |> sample_n(20000), aes(x=Tier,y=Rate,fill=Tier)) +
  geom_boxplot(width=0.5,outlier.size=0.3,outlier.alpha=0.1) +
  scale_fill_brewer(palette="Set2") + theme_minimal(base_size=8) +
  theme(legend.position="none") + labs(title="(d) Rate by Tier")

p_spread <- ecar |> group_by(Year) |>
  summarise(m=mean(Spread),se=sd(Spread)/sqrt(n()),.groups="drop") |>
  filter(Year>=2002,Year<=2012) |>
  ggplot(aes(x=Year,y=m)) +
  geom_ribbon(aes(ymin=m-1.96*se,ymax=m+1.96*se),fill="#1565C0",alpha=0.15) +
  geom_line(color="#1565C0",linewidth=1) + geom_point(color="#1565C0",size=2) +
  theme_minimal(base_size=8) + labs(title="(e) Spread Over Time",y="Spread (pp)")

p_car <- ggplot(ecar |> sample_n(20000), aes(x=CarType,y=Rate,fill=CarType)) +
  geom_violin(alpha=0.3) + geom_boxplot(width=0.1,outlier.shape=NA,fill="white") +
  scale_fill_manual(values=c(N="#1565C0",U="#E53935")) +
  theme_minimal(base_size=8) + theme(legend.position="none") + labs(title="(f) New vs Used")

dashboard <- (p_fico|p_rate|p_scatter) / (p_tier|p_spread|p_car) +
  plot_annotation(title="e-Car Loan: EDA Dashboard (208K applications)",
    caption="Source: e-Car Dataset | UNYT", tag_levels="a") &
  theme_minimal(base_size=8)
ggsave("output/W04_M08_dashboard.pdf",dashboard,width=14,height=9)
ggsave("output/W04_M08_dashboard.png",dashboard,width=14,height=9,dpi=300)

# ── STRUCTURAL MNAR ──
p_shadow <- ecar |> mutate(PrevAvail=ifelse(is_new,"No Prev Rate","Has Prev Rate")) |>
  ggplot(aes(x=FICO,fill=PrevAvail)) +
  geom_histogram(bins=40,position="identity",alpha=0.5) +
  scale_fill_manual(values=c("Has Prev Rate"="#1565C0","No Prev Rate"="#E53935")) +
  theme_minimal() + labs(title="FICO by Previous Rate Availability (MNAR)")
ggsave("output/W04_M08_shadow.png",p_shadow,width=7,height=4,dpi=300)

# ── KEY STATS ──
cat("Spearman(FICO,Rate):", round(cor(ecar$FICO,ecar$Rate,method="spearman"),3), "\n")
cat("Previous Rate missing:", round(mean(ecar$is_new)*100,1), "%\n")
cat("Median Rate (New):", round(median(ecar$Rate[ecar$CarType=="N"]),2), "%\n")
cat("Median Rate (Used):", round(median(ecar$Rate[ecar$CarType=="U"]),2), "%\n")
cat("\n── All W04-M08 R outputs saved ──\n")

File: latest/week2/workshop4_m8/W04_M08_demo.py
"""W04-M08: EDA Case Study — e-Car Loan Pricing — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.gridspec import GridSpec
from scipy.stats import spearmanr
os.makedirs("output",exist_ok=True); np.random.seed(42)

df=pd.read_csv("ecar.csv"); df.columns=[c.strip().replace("  "," ") for c in df.columns]
df["Car Type"]=df["Car Type"].str.strip()
df["Approve Date"]=pd.to_datetime(df["Approve Date"],format="%m/%d/%Y",errors="coerce")
df["Year"]=df["Approve Date"].dt.year
df["Previous Rate"]=pd.to_numeric(df["Previous Rate"],errors="coerce")
df["Spread"]=df["Rate"]-df["Cost of Funds"]
df["is_new"]=df["Previous Rate"].isna()
print(f"Rows: {len(df)}, Cols: {df.shape[1]}")

# Dashboard (6 panels)
fig=plt.figure(figsize=(14,9)); gs=GridSpec(2,3,figure=fig,hspace=0.4,wspace=0.35)
sub=df.sample(20000,random_state=42)

ax=fig.add_subplot(gs[0,0]); ax.hist(sub["FICO"],bins=50,color="#1565C0",edgecolor="white",lw=0.3)
ax.set_title("(a) FICO",fontsize=9,fontweight="bold")
ax=fig.add_subplot(gs[0,1]); ax.hist(sub["Rate"],bins=50,color="#2E7D32",edgecolor="white",lw=0.3)
ax.set_title("(b) Rate",fontsize=9,fontweight="bold")
ax=fig.add_subplot(gs[0,2]); s2=sub.sample(5000); ax.scatter(s2["FICO"],s2["Rate"],s=2,c="#1565C0",alpha=0.08)
rho,_=spearmanr(s2["FICO"],s2["Rate"]); ax.set_title(f"(c) FICO vs Rate (\u03C1={rho:.2f})",fontsize=9,fontweight="bold")
ax=fig.add_subplot(gs[1,0]); tiers=sorted(sub["Tier"].dropna().unique())
bp=ax.boxplot([sub.query("Tier==@t")["Rate"] for t in tiers],labels=[f"T{int(t)}" for t in tiers],patch_artist=True,widths=0.5)
cols5=["#1565C0","#2E7D32","#E65100","#7B1FA2","#C62828"]
for p,c in zip(bp["boxes"],cols5): p.set_facecolor(c); p.set_alpha(0.3)
ax.set_title("(d) Rate by Tier",fontsize=9,fontweight="bold")
ax=fig.add_subplot(gs[1,1]); yr=sub.groupby("Year")["Spread"].agg(["mean","std","count"])
yr["se"]=yr["std"]/np.sqrt(yr["count"]); yr=yr.loc[2002:2012]
ax.plot(yr.index,yr["mean"],"-o",color="#1565C0",lw=1.5,markersize=4)
ax.fill_between(yr.index,yr["mean"]-1.96*yr["se"],yr["mean"]+1.96*yr["se"],alpha=0.15,color="#1565C0")
ax.set_title("(e) Spread Over Time",fontsize=9,fontweight="bold")
ax=fig.add_subplot(gs[1,2])
for label,q,c in [("New","N","#1565C0"),("Used","U","#E53935")]:
    d=sub.query("`Car Type`==@q")["Rate"]; ax.hist(d,bins=40,alpha=0.4,color=c,label=f"{label} (med={d.median():.1f})")
ax.legend(fontsize=6); ax.set_title("(f) New vs Used",fontsize=9,fontweight="bold")

for a in fig.get_axes(): a.tick_params(labelsize=5); a.spines["top"].set_visible(False); a.spines["right"].set_visible(False)
fig.suptitle(f"e-Car Loan EDA: {len(df):,} applications",fontsize=13,fontweight="bold")
plt.savefig("output/W04_M08_dashboard.pdf",bbox_inches="tight")
plt.savefig("output/W04_M08_dashboard.png",dpi=300,bbox_inches="tight"); plt.close()

rho2,_=spearmanr(df["FICO"],df["Rate"])
print(f"Spearman(FICO,Rate)={rho2:.3f}")
print(f"Previous Rate missing: {df['is_new'].mean()*100:.1f}%")
print("All W04-M08 Python outputs saved")
